List availabale GDAL vector drivers (fix #18738)

alexbruy opened the pull request:
## Description
In QGIS settings there is a GDAL tab with list of available GDAL raster drivers and their capabilities. If several drivers can be used to open data, user can disable some drivers to make sure that data will be opened with required driver. Similar situation with vector formats. For example there are FileGDB and OpenFileGDB drivers to work with .gdb files, but if data file has version 9.x, FileGDB won't be able to open it. But there are no UI with list of available vector drivers.

This pull-requst adds "Vector Drivers" tab to the GDAL section in the QGIS options dialog. At this tab user can see all available vector drivers and their capabilities as well as uncheck drivers which should not be used to open vector data.

Raster drivers
![raster](https://user-images.githubusercontent.com/776954/91309205-fba33f80-e7b8-11ea-9d30-72e07ade781f.png)

Vector drivers
![vector](https://user-images.githubusercontent.com/776954/91309225-01992080-e7b9-11ea-8350-ce9ea2821caa.png)

Fixes #18738.

alexbruy committed:
add list of available GDAL vector drivers to the QGIS options (fix #18738)

This allows to select which driver to use if more than one driver can be
used to open dataset, as in the case of FileGDB.

alexbruy committed:
use comma as delimiter for GDAL_SKIP env variable as some vector drivers
have space in their name

SrNetoChan commented:
@alexbruy Once this PR gets merged, please assign the documentation ticket to me :-) thanks

nyalldawson commented on src/app/qgsoptions.cpp:
```diff
@@ -2259,6 +2259,7 @@ void QgsOptions::loadGdalDriverList()
   QStringList myDrivers;
   QStringList myGdalWriteDrivers;
   QMap<QString, QString> myDriversFlags, myDriversExt, myDriversLongName;
+  QMap<QString, bool> driversType; // true for raster, false for vector
```
Would be slightly nicer if this were QMap< QString, QgsMapLayerType >, but that's not a blocker!

alexbruy replied:
Changed to use `QgsMapLayerType`.

nyalldawson commented on src/app/qgsoptions.cpp:
```diff
@@ -2396,6 +2429,31 @@ void QgsOptions::saveGdalDriverList()
   }
+
+  // vector drivers
+  for ( int i = 0; i < lstVectorDrivers->topLevelItemCount(); i++ )
```
you could put the existing raster loop in a lambda and then re-use it here to avoid the duplicate code

alexbruy replied:
Should be fixed now. But not sure I did everything correctly.

nyalldawson commented on src/core/qgsapplication.cpp:
```diff
@@ -1687,7 +1687,7 @@ void QgsApplication::registerGdalDriversFromSettings()
   if ( !joinedList.isEmpty() )
   {
-    myList = joinedList.split( ' ' );
+    myList = joinedList.split( QStringLiteral( "," ) );
```
This will break existing configuration, right? Probably safer to change the setting key name, and use the old value with the space delimiter only if the newer (comma separated) key doesn't exist

alexbruy replied:
Yes, you are right. Fixed according to your suggestion.

nyalldawson commented on src/core/qgsapplication.cpp:
```diff
@@ -1707,7 +1707,7 @@ void QgsApplication::applyGdalSkippedDrivers()
       realDisabledDriverList << driverName;
   }
-  QString myDriverList = realDisabledDriverList.join( ' ' );
+  QString myDriverList = realDisabledDriverList.join( QStringLiteral( "," ) );
```
```suggestion
  QString myDriverList = realDisabledDriverList.join( ',' );
```

alexbruy replied:
Fixed.

nyalldawson commented:
Great work -- this will be very useful!!

alexbruy committed:
use layer type enum instead of bool flags
fix use of QStringLiteral

alexbruy committed:
move duplicated code to lambda

alexbruy committed:
improve driver type detection

alexbruy committed:
don't break existing configuration

Introduce new skipDrivers setting to store comma separated list of
disabled drivers. If this new setting does not exists in the QGIS
settings then disabled drivers will be loaded from old skipList setting.
